nilmtk/api.py

```diff
@@ -208,36 +209,93 @@ def train_jointly(self,clf,d):
-                    train_df, appliance_readings = self.dropna(train_df, appliance_readings)
+                if self.use_activations:
+                    print("Getting appliance activations")
+                    for appliance_name in self.appliances:
+                        train.set_window(start=d[dataset]['buildings'][building]['start_time'],
+                                         end=d[dataset]['buildings'][building]['end_time'])
+
+                        activations = train.buildings[building].elec[appliance_name].activation_series(
+                                self.appliances_params[appliance_name]["min_off_time"], 
+                                self.appliances_params[appliance_name]["min_on_time"], 
+                                self.appliances_params[appliance_name]["number_of_activation_padding"], 
+                                self.appliances_params[appliance_name]["min_on_power"])
+                            
+                        for period in activations:
+                            timestamp = period.index[0]
+                            tz_offset = (timestamp.replace(tzinfo=None) - 
+                                        timestamp.tz_convert('UTC').replace(tzinfo=None))
+                            tz_offset_bg = np.timedelta64(tz_offset)
+
+                            timestamp = period.index[-1]
+                            tz_offset = (timestamp.replace(tzinfo=None) - 
+                                        timestamp.tz_convert('UTC').replace(tzinfo=None))
+
+                            tz_offset_end = np.timedelta64(tz_offset)
+
+                            if tz_offset_bg > tz_offset_end:
+                                beg = period.index[0].tz_localize(None) - tz_offset_bg
+                                end = period.index[-1].tz_localize(None)
+                            elif tz_offset_bg < tz_offset_end: 
+                                beg = period.index[0].tz_localize(None)
+                                end = period.index[-1].tz_localize(None) - tz_offset_end 
+                            else:
+                                beg = period.index[0].tz_localize(None)
+                                end = period.index[-1].tz_localize(None)
+
+                            train.set_window(
+                                    start=beg,
+                                    end=end
+                                    )
+
+                            mains_df = next(train.buildings[building].elec.mains().load(physical_quantity='power', ac_type=self.power['mains'], sample_period=self.sample_period))
+                            
+                            if not mains_df.empty:
+                                mains_df = mains_df[[list(mains_df.columns)[0]]]
+
+                                appliance_df = next(train.buildings[building].elec[appliance_name].load(physical_quantity='power', ac_type=self.power['appliance'], sample_period=self.sample_period))
```

Comment:
What you put in `appliance_df` here is the same data as in your `period` variable. Why loading the data once more?

Reply:
I believed that the period variable only contained the beginning and end of the activation, not the actual data.